### ISSUE TITLE
update-all command throw exception : Trying to get property 'db' of non-object

### ISSUE THREAD

**Issue by samuelbbtech** [bizley/yii2-migration#33]
i use composer to install this library, everything ok, i can list, create, create-all but i cannot use update-all, it will prompt me "Trying to get property 'db' of non-object"

**Comment by bizley** [comment]
Could you provide more specific information? Stack trace, log?

**Comment by bizley** [comment]
Since there is no response I'm closing this. Feel free to reopen it if needed.

**Comment by grinvich11** [comment]
Hi, i have the same case. It is reproduced using Postgres database. If you use Mysql everything is ok.
1. migration/create table_name
2. add column to table
3. migration/update table_name
```
Exception 'Error' with message 'Call to a member function createQueryBuilder() on null' 
/vendor/bizley/migration/src/dummy/Migration.php:89
```
PHP 7.1, Yii 2.0.15.1, package version - 3.0.5

**Comment by bizley** [comment]
Could you provide your full console config?

**Comment by bizley** [comment]
And the created migration code please.

**Comment by grinvich11** [comment]
Created migration

```
<?php

use yii\db\Migration;

class m181107_135903_create_table_news extends Migration
{
    public function up()
    {
        $tableOptions = null;
        if ($this->db->driverName === 'mysql') {
            $tableOptions = 'CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci ENGINE=InnoDB';
        }

        $this->createTable('{{%news}}', [
            'id' => $this->smallInteger()->notNull()->append('PRIMARY KEY'),
            'name' => $this->string(),
            'published' => $this->smallInteger()->notNull()->defaultValue('0'),
        ], $tableOptions);

    }

    public function down()
    {
        $this->dropTable('{{%news}}');
    }
}
```

Stack trace after adding a new column to the table
```
yii migration/update news

Yii 2 Migration Generator Tool v3.0.5

 > Generating update migration for table 'news' ...Exception 'Error' with message 'Call to a member function createQueryBuilder() on null'

in \project\vendor\bizley\migration\src\dummy\Migration.php:89

Stack trace:
#0 \project\vendor\bizley\migration\src\dummy\Migration.php(149): yii\db\Migration->getKeysMap(Object(yii\db\ColumnSchemaBuilder))
#1 \project\vendor\bizley\migration\src\dummy\Migration.php(77): yii\db\Migration->extractColumn(Object(yii\db\ColumnSchemaBuilder))
#2 \project\vendor\bizley\migration\src\dummy\Migration.php(238): yii\db\Migration->extractColumns(Array)
#3 \project\console\migrations\m181107_135903_create_table_news.php(18): yii\db\Migration->createTable('{{%news}}', Array, NULL)
#4 \project\vendor\bizley\migration\src\Updater.php(166): m181107_135903_create_table_news->up()
#5 \project\vendor\bizley\migration\src\Updater.php(453): bizley\migration\Updater->extract('m181107_135903_...')
#6 \project\vendor\bizley\migration\src\controllers\MigrationController.php(464): bizley\migration\Updater->isUpdateRequired()
#7 [internal function]: bizley\migration\controllers\MigrationController->actionUpdate('news')
#8 \project\vendor\yiisoft\yii2\base\InlineAction.php(57): call_user_func_array(Array, Array)
#9 \project\vendor\yiisoft\yii2\base\Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#10 \project\vendor\yiisoft\yii2\console\Controller.php(148): yii\base\Controller->runAction('update', Array)
#11 \project\vendor\yiisoft\yii2\base\Module.php(528): yii\console\Controller->runAction('update', Array)
#12 \project\vendor\yiisoft\yii2\console\Application.php(180): yii\base\Module->runAction('migration/updat...', Array)
#13 \project\vendor\yiisoft\yii2\console\Application.php(147): yii\console\Application->runAction('migration/updat...', Array)
#14 \project\vendor\yiisoft\yii2\base\Application.php(386): yii\console\Application->handleRequest(Object(yii\console\Request))
#15 \project\yii(23): yii\base\Application->run()
#16 {main}
```

**Comment by bizley** [comment]
And the full console config as well, of course remove all sensitive data.

**Comment by grinvich11** [comment]
If I use the connection to the MySQL, then this problem does not occur
```
...
'components' => [
        'db' => [
            'class' => 'yii\db\Connection',
            'dsn' => 'pgsql:host=127.0.0.1;port=5432;dbname=test',
            'username' => 'postgres',
            'password' => '',
            'charset' => 'utf8',
        ],
],
'controllerMap' => [
        'migration' => [
            'class' => 'bizley\migration\controllers\MigrationController',
        ],
],
'modules' => [
        'gii' => 'yii\gii\Module',
],
...
```

**Comment by bizley** [comment]
Could you check if this helps?

Please change line 149 in `src/dummy/Migration.php`:

    $keyToDb = $this->getKeysMap($type);

to:

    $keyToDb = $this->db->schema->createQueryBuilder()->typeMap;

and run it again.

**Comment by grinvich11** [comment]
Got the following error

```
 > Generating update migration for table 'news' ...PHP Warning 'yii\base\ErrorException' with message 'array_key_exists() expects parameter 2 to be array, null given'

in \project\vendor\bizley\migration\src\dummy\Migration.php:132

Stack trace:
#0 \project\vendor\bizley\migration\src\dummy\Migration.php(156): m181107_135903_create_table_news->fillTypeMapProperties()
#1 \project\vendor\bizley\migration\src\dummy\Migration.php(77): m181107_135903_create_table_news->extractColumn()
#2 \project\vendor\bizley\migration\src\dummy\Migration.php(239): m181107_135903_create_table_news->extractColumns()
#3 \project\console\migrations\m181107_135903_create_table_news.php(18): m181107_135903_create_table_news->createTable()
#4 \project\vendor\bizley\migration\src\Updater.php(166): m181107_135903_create_table_news->up()
#5 \project\vendor\bizley\migration\src\Updater.php(453): bizley\migration\Updater->extract()
#6 \project\vendor\bizley\migration\src\controllers\MigrationController.php(464): bizley\migration\Updater->isUpdateRequired()
```

**Comment by bizley** [comment]
Sorry, forgot about one more. Line 150, from:

    $dbToKey = $type->db->schema->typeMap;

to

    $dbToKey = $this->db->schema->typeMap;

**Comment by grinvich11** [comment]
It helps!

**Comment by bizley** [comment]
Cheers! I'll prepare new version.

**Comment by bizley** [comment]
Fixed in [3.1.0](https://github.com/bizley/yii2-migration/releases/tag/3.1.0) / [2.4.0](https://github.com/bizley/yii2-migration/releases/tag/2.4.0)
